README.md
# 3dbarplot

R package for creating 3D barplots. Read the documentation or included vignette for full details.

Easily make beautiful 3D barplots like this:

<p align="center">
  <img src="man/figures/pretty.png">
</p>

Includes an easy-to-use wrapper function for making "legoplots" of mutation/trinucleotide context data like this:

<p align="center">
  <img src="man/figures/sig2sanger.png">
  <img src="man/figures/sig8broad.png">
</p>





NEWS.md
# barplot3d 1.0.1
* 2019/11 minor edits for CRAN resubmission.
# barplot3d 1.0.0
* 2019/10 first version submitted to CRAN.


R/bar3d.R
#' Adds a single 3D bar to the current scene
#'
#' @param x The x dimensions of the bar, a vector of length 2 e.g. c(0,1).
#' @param y The y dimensions of the bar, a vector of length 2 e.g. c(0,1).
#' @param z The height of the bar, a single number, e.g 3.
#' @param alpha The alpha channel (transparency) of the sides of the bar. Range 0-1.
#' @param topcol The color of the top of the bar. Text description or hexadecimal RGB color, like that returned by rgb() e.g. "red" or "#078E53"
#' @param sidecol The color of the sides of the bar. Text description or a hexadecimal RGB color, like that returned by rgb() e.g. "gray" or "#aaaaaa"
#' @param linecol The color of the edges of the bar. Text description or be a hexadecimal RGB color, like that returned by rgb() e.g. "black" or "#000000"
#' @return Nothing is returned (invisibly returns NULL).
#' @examples
#' \dontshow{
#' options(rgl.useNULL=TRUE)
#' }
#' bar3d(c(0,1),c(0,1),3,alpha=0.6,topcol="#078E53",sidecol="#aaaaaa",linecol="#000000")
#' @export
bar3d<-function(x=c(0,1),y=c(0,1),z,alpha=1,topcol="#078E53",sidecol="#aaaaaa",linecol="#000000"){

  ## These lines allow the active rgl device to be updated with multiple changes
  ## This is necessary to draw the sides and ends of the column separately
  save <- rgl::par3d(skipRedraw=TRUE)
  on.exit(rgl::par3d(save))

  ## Determine the coordinates of each surface of the column and its edges
  x1=c(rep(c(x[1],x[2],x[2],x[1]),3),rep(x[1],4),rep(x[2],4))
  z1=c(rep(0,4),rep(c(0,0,z,z),4))
  y1=c(y[1],y[1],y[2],y[2],rep(y[1],4),rep(y[2],4),rep(c(y[1],y[2],y[2],y[1]),2))
  x2=c(rep(c(x[1],x[1],x[2],x[2]),2),rep(c(x[1],x[2],rep(x[1],3),rep(x[2],3)),2))
  z2=c(rep(c(0,z),4),rep(0,8),rep(z,8) )
  y2=c(rep(y[1],4),rep(y[2],4),rep(c(rep(y[1],3),rep(y[2],3),y[1],y[2]),2) )

  ## These lines create the sides of the column and its colored top surface
  rgl::rgl.quads(x1,z1,y1,col=rep(sidecol,each=4),alpha=alpha,lit=FALSE)
  rgl::rgl.quads(c(x[1],x[2],x[2],x[1]),rep(z,4),c(y[1],y[1],y[2],y[2]),
            col=rep(topcol,each=4),alpha=1,lit=FALSE)
  ## Add colored edges to the column
  rgl::rgl.lines(x2,z2,y2,col=linecol,lit=FALSE)
  invisible()
}


R/barplot3d.R
#' Adds a 3D bar plot to the current RGL scene
#' @param rows How many rows the plotting area should have, an integer, e.g. 5.
#' @param cols How many columns the plotting area should have, an integer, e.g. 5.
#' @param x The x dimensions of each 3D bar, a vector of length 2 e.g. c(0,1).
#' @param y The y dimensions of each 3D bar, a vector of length 2 e.g. c(0,1).
#' @param z The height of each 3D bar, a numeric vector, e.g c(2,3,5,2,9).
#' @param alpha The alpha channel (transparency) of the sides of 3D bars. Range 0-1.
#' @param scalexy Scaling factor for x and y coordinates; this constant can be used to make the plot "skinnier" or "fatter".
#' @param gap Gap between 3D bars (recommended values are 0 or 0.2).
#' @param topcolors The color of the top of each 3D bar. Numeric vector of hexadecimal RGB colors, like those returned by rgb() e.g. "#078E53".
#' @param sidecolors The color of the top of the bar. Should be a hexadecimal RGB color, like that returned by rgb() e.g. "#aaaaaa".
#' @param linecolors The color of the edges of the bar. Should be a hexadecimal RGB color, like that returned by rgb() e.g. "#aaaaaa".
#' @param theta Polar coordinate for viewing the 3D barplot; range 0 to 360 (rotates the plot).
#' @param phi Polar coordinate for viewing the 3D barplot; range -90 to 90 (-90 is directly below, 90 directly above).
#' @param gridlines Draw gridlines on the plot (TRUE or FALSE).
#' @param xlabels Labels for the x axis (must be a vector of names the same length as "cols" parameter).
#' @param ylabels Labels for the y axis (must be a vector of names the same length as "rows" parameter).
#' @param zlabels Labels for the z axis; add numeric scale to the vertical dimension of the plot (TRUE or FALSE).
#' @param xsub Descriptive label for the x axis.
#' @param ysub Descriptive label for the y axis.
#' @param zsub Descriptive label for the z axis.
#' @return Nothing is returned (invisibly returns NULL).
#' @examples
#' \dontshow{
#' options(rgl.useNULL=TRUE)
#' }
#' barplot3d(rows=3,cols=5,z=1:12,topcolors=rainbow(12),alpha=0.7,scalexy=10,
#' xlabels=c("One","Two","Three","Four","Five"),ylabels=LETTERS[1:3])
#' @export
barplot3d<-function(rows,cols,x,y,z,alpha=1,scalexy=1,gap=0.2,topcolors=c("#000000"),sidecolors=c("#aaaaaa"),linecolors=c("#000000"),
                    theta=50,phi=40,gridlines=TRUE,xlabels=FALSE,ylabels=FALSE,zlabels=TRUE,xsub=FALSE,ysub=FALSE,zsub=FALSE){

  ## Test to ensure that the data matches the dimensions given
  if(length(z)>(rows*cols)){
    stop("Too many data points for dimensions of plot. z must be equal to or less than rows*cols", call. = FALSE)
  }
  ## Test that (if provided) the xlabels match the length of the cols parameter
  if(xlabels[1]!=FALSE){
    if(length(xlabels)!=cols){
      stop("Number of xlabels must be equal to the cols parameter", call. = FALSE)
    }
  }
  ## Test that (if provided) the ylabels match the length of the rows parameter
  if(ylabels[1]!=FALSE){
    if(length(ylabels)!=rows){
      stop("Number of ylabels must be equal to the rows parameter", call. = FALSE)
    }
  }
  ## If only one topcolor is given, replicate it for all bars
  if(length(topcolors==1)){
    topcolors=rep(topcolors,length(z))
  }
  ## If only one sidecolor is given, replicate it for all bars
  if(length(sidecolors==1)){
    sidecolors=rep(sidecolors,length(z))
  }
  ## If only one linecolor is given, replicate it for all bars
  if(length(linecolors==1)){
    linecolors=rep(linecolors,length(z))
  }

  ## These lines allow the active rgl device to be updated with multiple changes
  ## This is necessary to add each column sequentially
  save <- rgl::par3d(skipRedraw=TRUE)
  on.exit(rgl::par3d(save))

  ## Define dimensions of the plot
  dimensions=c(rows,cols)

  ## Scale column area and the gap between columns
  y=seq(1,dimensions[1])*scalexy
  x=seq(1,dimensions[2])*scalexy
  gap=gap*scalexy

  ## Plot each of the columns
  for(i in 1:dimensions[1]){
    for(j in 1:dimensions[2]){
      it=(i-1)*dimensions[2]+j # Variable to work out which column to plot; counts from 1:length(z)
      if(it>length(z)){next} # Do not plot beyond the length of the input
      bar3d(c(gap+x[j],x[j]+scalexy),
            c(-gap-y[i],-y[i]-scalexy),
            z[it],
            alpha=alpha,
            topcol=topcolors[it],
            sidecol=sidecolors[it],
            linecol=linecolors[it])
    }
  }
  ## Set the viewpoint
  rgl::rgl.viewpoint(theta=theta,phi=phi,fov=0)

  ## If desired, add grid lines, axes labels and axes subtitles
  if(gridlines){
    rgl::grid3d("y",col="black",at=((1:cols)+0.4)*scalexy+gap) # this is the floor and matches the "cols" parameter
    rgl::grid3d("x",col="black") # this is the wall and matches the z vertical direction
    rgl::grid3d("z",col="black") # this is the wall and matches the z vertical direction
    rgl::grid3d("y",col="black",at=(((1:rows)+0.4)*scalexy+gap)*-1) # this is the floor and matches the "rows" parameter
  }
  if(xlabels[1]!=FALSE){
    rgl::axis3d("x-+", labels=xlabels,at=((1:cols)+0.4)*scalexy+gap) # this is the floor and matches the "cols" parameter
  }
  if(ylabels[1]!=FALSE){
    rgl::axis3d("z+", labels=ylabels,at=-1*((1:rows)+0.4)*scalexy-gap) # this is the floor and matches the "rows" parameter
  }
  if(zlabels){
    rgl::axis3d("y-+", labels=TRUE) # vertical numeric scale
  }
  if(xsub!=FALSE){
    rgl::mtext3d(xsub,edge="x-+",line=2)
  }
  if(ysub!=FALSE){
    rgl::mtext3d(ysub,edge="z+",line=2)
  }
  if(zsub!=FALSE){
    rgl::mtext3d(zsub,edge="y-+",line=2)
  }

  invisible()
}



R/singlecolumn.R
## Function modified from the "demo(hist3d)" examples in the rgl package:
# library(rgl)
# demo(hist3d)

## Draws a single "column" or "stack".
## X and Y coordinates determine the area of the column
## The Z coordinate determines the height of the column
## We include "lit=FALSE" arguments to remove the nasty shiny surfaces caused by lighting
singlecolumn=function(x,y,z,alpha=1,topcol="#078E53",sidecol="#aaaaaa",linecol="#000000"){
  
  ## These lines allow the active rgl device to be updated with multiple changes
  ## This is necessary to draw the sides and ends of the column separately  
  save=par3d(skipRedraw=TRUE)
  on.exit(par3d(save))
  
  ## Determine the coordinates of each surface of the column and its edges
  x1=c(rep(c(x[1],x[2],x[2],x[1]),3),rep(x[1],4),rep(x[2],4))
  z1=c(rep(0,4),rep(c(0,0,z,z),4))
  y1=c(y[1],y[1],y[2],y[2],rep(y[1],4),rep(y[2],4),rep(c(y[1],y[2],y[2],y[1]),2))
  x2=c(rep(c(x[1],x[1],x[2],x[2]),2),rep(c(x[1],x[2],rep(x[1],3),rep(x[2],3)),2))
  z2=c(rep(c(0,z),4),rep(0,8),rep(z,8))
  y2=c(rep(y[1],4),rep(y[2],4),rep(c(rep(y[1],3),rep(y[2],3),y[1],y[2]),2))
  
  ## These lines create the sides of the column and its coloured top surface
  rgl.quads(x1,z1,y1,col=rep(sidecol,each=4),alpha=alpha,lit=FALSE)
  rgl.quads(c(x[1],x[2],x[2],x[1]),rep(z,4),c(y[1],y[1],y[2],y[2]),
            col=rep(topcol,each=4),alpha=1,lit=FALSE) 
  ## This line adds black edges to the column
  rgl.lines(x2,z2,y2,col=linecol,lit=FALSE)
}
# Example:
# stackplot.3d(c(0,1),c(0,1),3,alpha=0.6)

R/legoplot.R
#' A wrapper function to create a sequence context "legoplot"
#' @param contextdata A numeric vector of counts or frequencies of the 96 possible somatic mutations and trinucleotide contexts. These MUST be in the same order as in the example (see example and/or vignette).
#' @param alpha The alpha channel (transparency) of the sides of 3D bars. Range 0-1.
#' @param scalexy Scaling factor for x and y coordinates; this constant can be used to make the plot "skinnier" or "fatter".
#' @param gap Gap between 3D bars.
#' @param sixcolors The color scheme. "broad" for Broad Institute colors, "sanger" for Sanger Institute colors or a vector of six hexadecimal RGB colors.
#' @param theta Polar coordinate for viewing the 3D barplot; range 0 to 360 (rotates the plot).
#' @param phi Polar coordinate for viewing the 3D barplot; range -90 to 90 (-90 is directly below, 90 directly above).
#' @param gridlines Draw gridlines on the plot (TRUE or FALSE).
#' @param labels Include the default axis labels (TRUE or FALSE).
#' @param zlabels Labels for the z axis; add numeric scale to the vertical dimension of the plot (TRUE or FALSE).
#' @param zsub Descriptive label for the z axis.
#' @return Nothing is returned (invisibly returns NULL).
#' @examples
#' \dontshow{
#' options(rgl.useNULL=TRUE)
#' }
#' # Read in COSMIC signature probabilities
#' x=system.file("extdata", "signature_probabilities.txt", package = "barplot3d")
#' sigdata=read.table(x,header=TRUE,stringsAsFactors = FALSE)
#' # Plot signature 2 with Sanger colors and some transparency so we can see all bars
#' legoplot3d(contextdata=sigdata$Signature_2,labels=TRUE,scalexy=0.05,sixcolors="sanger",
#' alpha=0.4,zsub="Probability")
#' @export
legoplot3d<-function(contextdata,alpha=1,scalexy=1,gap=0.2,sixcolors="broad",
                    theta=50,phi=40,gridlines=TRUE,labels=FALSE,zlabels=TRUE,zsub=FALSE){

  ## Setup the 6 color regions
  set1=c(1:4,17:20,5:8,21:24,9:12,25:28,13:16,29:32)
  set2=set1+32
  set3=set1+64
  neworder=c(set1,set2,set3)
  if(sixcolors[1]=="broad"){
    sixcolors=c("#805D3F","#72549A","#5EAFB2","#3F4F9D","#F2EC3C","#74B655")
  }
  if(sixcolors[1]=="sanger"){
    sixcolors=c("#050708","#CBCACB","#1EBFF0","#EDC8C5","#E62725","#A1CF64")
  }
  mycolors=as.vector(sapply(sixcolors,rep,16))[neworder]

  ## Set label variables
  if(labels){
    xlabels=rep(c("A","C","G","T"),2)
    ylabels=rep(c("A","C","G","T"),3)
    xsub="Preceeding base"
    ysub="Succeeding base"
  }else{
    xlabels=FALSE
    ylabels=FALSE
    xsub=""
    ysub=""
  }

  ## Reorder input data to plotting order
  porder=order(c(33,34,35,36,41,42,43,44,49,50,51,52,57,58,59,60,1,2,3,4,9,10,11,12,17,18,19,20,25,26,27,28,65,66,67,68,73,74,75,76,81,82,83,84,89,90,91,92,5,6,7,8,13,14,15,16,21,22,23,24,29,30,31,32,69,70,71,72,77,78,79,80,85,86,87,88,93,94,95,96,37,38,39,40,45,46,47,48,53,54,55,56,61,62,63,64))

  barplot3d(rows=12,cols=8,z=contextdata[porder],alpha=alpha,scalexy=scalexy,gap=gap,topcolors=mycolors,sidecolors=mycolors,
                      theta=50,phi=40,gridlines=TRUE,xlabels=xlabels,ylabels=ylabels,zlabels=zlabels,xsub=xsub,ysub=ysub,zsub=zsub)
}
